cpp/src/c_api/renumber_arbitrary_edgelist.cu

```diff
@@ -0,0 +1,215 @@
+        vertex_t old_vertex_id = renumber_chunk_span[idx];
+        vertex_t new_vertex_id = static_cast<vertex_t>(chunk_base_offset + idx);
+
+        auto begin_iter =
+          thrust::lower_bound(thrust::seq, srcs_span.begin(), srcs_span.end(), old_vertex_id);
+        if (begin_iter != srcs_span.end()) {
+          auto end_iter =
+            thrust::upper_bound(thrust::seq, srcs_span.begin(), srcs_span.end(), old_vertex_id);
+
+          while (begin_iter != end_iter) {
+            size_t offset = thrust::distance(srcs_span.begin(), begin_iter);
+            output_srcs_span[srcs_pos_span[offset]] = new_vertex_id;
+            srcs_span[offset]                       = cugraph::invalid_vertex_id<vertex_t>();
+            ++begin_iter;
+          }
+        }
+
+        begin_iter =
+          thrust::lower_bound(thrust::seq, dsts_span.begin(), dsts_span.end(), old_vertex_id);
+        if (begin_iter != dsts_span.end()) {
+          auto end_iter =
+            thrust::upper_bound(thrust::seq, dsts_span.begin(), dsts_span.end(), old_vertex_id);
+
+          while (begin_iter != end_iter) {
+            size_t offset = thrust::distance(dsts_span.begin(), begin_iter);
+            output_dsts_span[dsts_pos_span[offset]] = new_vertex_id;
+            dsts_span[offset]                       = cugraph::invalid_vertex_id<vertex_t>();
+            ++begin_iter;
+          }
+        }
+      });
+
+    srcs_v.resize(thrust::distance(
+                    thrust::make_zip_iterator(srcs_v.begin(), srcs_pos.begin()),
+                    thrust::remove_if(handle.get_thrust_policy(),
+                                      thrust::make_zip_iterator(srcs_v.begin(), srcs_pos.begin()),
+                                      thrust::make_zip_iterator(srcs_v.end(), srcs_pos.end()),
+                                      [] __device__(auto t) {
+                                        return thrust::get<0>(t) ==
+                                               cugraph::invalid_vertex_id<vertex_t>();
+                                      })),
+                  handle.get_stream());
+    srcs_pos.resize(srcs_v.size(), handle.get_stream());
+
+    dsts_v.resize(thrust::distance(
+                    thrust::make_zip_iterator(dsts_v.begin(), dsts_pos.begin()),
+                    thrust::remove_if(handle.get_thrust_policy(),
+                                      thrust::make_zip_iterator(dsts_v.begin(), dsts_pos.begin()),
+                                      thrust::make_zip_iterator(dsts_v.end(), dsts_pos.end()),
+                                      [] __device__(auto t) {
+                                        return thrust::get<0>(t) ==
+                                               cugraph::invalid_vertex_id<vertex_t>();
+                                      })),
+                  handle.get_stream());
+    dsts_pos.resize(dsts_v.size(), handle.get_stream());
```

Comment:
I feel like scanning src_v & dst_v multiple times sounds a bit inefficient.

We already sorted srcs & dsts, so it is easy to find unique srcs & dsts, and accordingly unique vertices. Then, for each chunk, we can call thrust::copy_if to find only the relevant (old_v, new_v) pairs. If we have this renumber map, we can just directly renumber so no need for position vectors as well.

Reply:
Refactored based on this guidance.